helper/keysutil/lock_manager.go

```diff
@@ -288,112 +143,156 @@ func (lm *LockManager) RestorePolicy(ctx context.Context, storage logical.Storag
 		return errwrap.Wrapf(fmt.Sprintf("failed to restore the policy %q: {{err}}", name), err)
 	}
 
+	keyData.Policy.l = new(sync.RWMutex)
+
 	// Update the cache to contain the restored policy
-	lm.UpdateCache(name, keyData.Policy)
+	lm.cache.Store(name, keyData.Policy)
 
 	return nil
 }
 
 func (lm *LockManager) BackupPolicy(ctx context.Context, storage logical.Storage, name string) (string, error) {
-	p, lock, err := lm.GetPolicyExclusive(ctx, storage, name)
-	if lock != nil {
-		defer lock.Unlock()
-	}
-	if err != nil {
-		return "", err
+	var p *Policy
+	var err error
+
+	// Backup writes information about when the bacup took place, so we get an
+	// exclusive lock here
+	lock := locksutil.LockForKey(lm.keyLocks, name)
+	lock.Lock()
+	defer lock.Unlock()
+
+	pRaw, ok := lm.cache.Load(name)
+	if ok {
+		p = pRaw.(*Policy)
+		p.l.Lock()
+		defer p.l.Unlock()
+	} else {
+		// If the policy doesn't exit in storage, error out
+		p, err = lm.getPolicyFromStorage(ctx, storage, name)
+		if err != nil {
+			return "", err
+		}
+		if p == nil {
+			return "", fmt.Errorf(fmt.Sprintf("key %q not found", name))
+		}
 	}
-	if p == nil {
-		return "", fmt.Errorf("invalid key %q", name)
+
+	if atomic.LoadUint32(&p.deleted) == 1 {
+		return "", fmt.Errorf(fmt.Sprintf("key %q not found", name))
 	}
 
 	backup, err := p.Backup(ctx, storage)
 	if err != nil {
 		return "", err
 	}
 
-	// Update the cache since the policy would now have the backup information
-	lm.UpdateCache(name, p)
-
 	return backup, nil
 }
 
-// When the function returns, a lock will be held on the policy if err == nil.
-// It is the caller's responsibility to unlock.
-func (lm *LockManager) getPolicyCommon(ctx context.Context, req PolicyRequest, lockType bool) (*Policy, *sync.RWMutex, bool, error) {
-	lock := lm.policyLock(req.Name, lockType)
-
+// When the function returns, if caching was disabled, the Policy's lock must
+// be unlocked when the caller is done (and it should not be re-locked).
+func (lm *LockManager) GetPolicy(ctx context.Context, req PolicyRequest) (*Policy, bool, error) {
 	var p *Policy
 	var err error
 
 	// Check if it's in our cache. If so, return right away.
-	if lm.CacheActive() {
-		lm.cacheMutex.RLock()
-		p = lm.cache[req.Name]
-		if p != nil {
-			lm.cacheMutex.RUnlock()
-			return p, lock, false, nil
+	pRaw, ok := lm.cache.Load(req.Name)
+	if ok {
+		p = pRaw.(*Policy)
+		if atomic.LoadUint32(&p.deleted) == 1 {
+			return nil, false, nil
+		}
+		return p, false, nil
+	}
+
+	// We're not using the cache, or it wasn't found; get an exclusive lock.
+	// This ensures that any other process writing the actual storage will be
+	// finished before we load from storage.
+	lock := locksutil.LockForKey(lm.keyLocks, req.Name)
+	lock.Lock()
+
+	// If we are using the cache, defer the lock unlock; otherwise we will
+	// return from here with the lock still held.
+	cleanup := func() {}
+	if lm.useCache {
+		cleanup = func() {
+			lock.Unlock()
```

Comment:
The complexity here of unlocking sometimes via a cleanup func that may or may not have any contents seems a little tricky to me. I wonder if there might be a simpler way that would keep the unlocking a little more encapsulated. Like, I'm not sure if it would work, but maybe the cleanup func would _always_ be set, and it would be something more like, if using a cache unlock it, otherwise replace this lock with a new lock.

Reply:
And then that func would just be deferred near the top and called once.